python/tvm/runtime/disco/session.py

```diff
@@ -0,0 +1,294 @@
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""This module defines a Session in Disco. Session is the primary interface that users interact
+with the distributed runtime.
+"""
+from typing import Any, Callable, Optional, Sequence
+
+from ..._ffi import register_object
+from ..._ffi.runtime_ctypes import Device
+from ..ndarray import NDArray
+from ..object import Object
+from . import _ffi_api
+
+
+@register_object("runtime.disco.DRef")
+class DRef(Object):
+    """An object that exists on all workers. The controller process assigns a unique "register id"
+    to each object, and the worker process uses this id to refer to the object residing on itself.
+    """
+
+    @property
+    def session(self) -> "Session":
+        """Get the session that this DRef belongs to."""
+        return _ffi_api.DRefSession(self)  # type: ignore # pylint: disable=no-member
+
+    def debug_get_from_remote(self, worker_id: int) -> Any:
+        """Get the value of a DRef from a remote worker.
+
+        Parameters
+        ----------
+        worker_id : int
+            The id of the worker to be fetched from.
+
+        Returns
+        -------
+        value : object
+            The value of the register.
+        """
+        return _ffi_api.DRefDebugGetFromRemote(self, worker_id)  # type: ignore # pylint: disable=no-member
+
+
+class DPackedFunc(DRef):
+    """A PackedFunc in a Disco session."""
+
+    def __init__(self, dref: DRef) -> None:
+        self.handle = dref.handle
+        dref.handle = None
+
+    def __call__(self, *args) -> DRef:
+        return self.session.call_packed(self, *args)
+
+
+class DModule(DRef):
+    """A Module in a Disco session."""
+
+    def __init__(self, dref: DRef) -> None:
+        self.handle = dref.handle
+        del dref.handle
+
+    def __getitem__(self, name: str) -> DPackedFunc:
+        func = self.session._get_cached_method("runtime.ModuleGetFunction")
+        return DPackedFunc(func(self, name, False))
+
+
+@register_object("runtime.disco.Session")
+class Session(Object):
+    """A Disco interactive session. It allows users to interact with the Disco command queue with
+    various PackedFunc calling convention."""
+
+    @staticmethod
+    def threaded_session(num_workers: int) -> "Session":
+        """Create a threaded session."""
```

Comment:
```suggestion
        """Create a thread-scope session."""
```
I don't understand what "threaded" means at first glance.
Maybe use a better name for threaded_session as well.

Reply:
to expand a bit

```
Create a disco session backed by multiple threads in the same process.
```